File: Attributor/Objective-C/Attributor/Attributor/TextStatsViewController.m
//
//  TextStatsViewController.m
//  Attributor
//
//  Created by Afonso Graça on 22/10/14.
//  Copyright (c) 2014 Afonso Graça. All rights reserved.
//

#import "TextStatsViewController.h"

@interface TextStatsViewController ()
@property (weak, nonatomic) IBOutlet UILabel *colourfulCharacterLabel;
@property (weak, nonatomic) IBOutlet UILabel *outlinedCharacterLabel;

@end

@implementation TextStatsViewController

//- (void)viewDidLoad {
//    [super viewDidLoad];
//    // Do any additional setup after loading the view.
//	self.textToAnalyse = [[NSAttributedString alloc] initWithString:@"test" attributes:@{NSForegroundColorAttributeName : [UIColor greenColor], NSStrokeWidthAttributeName : @-3}];
//}

- (void)viewWillAppear:(BOOL)animated {
	[super viewWillAppear:animated];
	[self updateUI];
}

- (void)updateUI {
	self.colourfulCharacterLabel.text = [NSString stringWithFormat:@"%lu colourful characters", (unsigned long)[[self charactersWithAttribute: NSForegroundColorAttributeName] length]];
	self.outlinedCharacterLabel.text = [NSString stringWithFormat:@"%lu outlined characters", (unsigned long)[[self charactersWithAttribute: NSStrokeWidthAttributeName] length]];
}

- (NSAttributedString *)charactersWithAttribute: (NSString *)attribute {
	NSMutableAttributedString *characters = [[NSMutableAttributedString alloc] init];
	
	int index = 0;
	
	while(index < [self.textToAnalyse length]){
		NSRange range;
		id value = [self.textToAnalyse attribute:attribute atIndex:index effectiveRange:&range];
		if (value) {
			[characters appendAttributedString:[self.textToAnalyse attributedSubstringFromRange:range]];
			index =  (int)(range.location + range.length);
		}
		else {
			index++;
		}
	}
	return characters;
}

@end

File: Attributor/Swift/Attributor/Attributor/TextStatsViewController.swift
//
//  TextStatsViewController.swift
//  Attributor
//
//  Created by Afonso Graça on 20/10/14.
//  Copyright (c) 2014 Afonso Graça. All rights reserved.
//

import UIKit

class TextStatsViewController : UIViewController {
	
//	override func viewDidLoad() {
//		super.viewDidLoad()
//		self.textToAnalyse = NSAttributedString(string: "test", attributes: [NSForegroundColorAttributeName : UIColor.greenColor(), NSStrokeWidthAttributeName : -3])
//	}
	
	var textToAnalyse : NSAttributedString = NSAttributedString() {
		didSet {
			if self.view.window != nil {self.updateUI()}
		}
	}
	
	@IBOutlet weak var colourfulCharacterLabel: UILabel!
	@IBOutlet weak var outlinedCharactersLabel: UILabel!
	
	override func viewWillAppear(animated: Bool) {
		super.viewWillAppear(animated)
		self.updateUI()
	}
	
	func updateUI() {
		self.colourfulCharacterLabel.text = "\(self.charactersWith(NSForegroundColorAttributeName).length) colourful characters"
		self.outlinedCharactersLabel.text = "\(self.charactersWith(NSStrokeWidthAttributeName).length) outlined characters"
	}
	
	func charactersWith(attribute : String) -> NSAttributedString {
		var characters = NSMutableAttributedString()
		
		var index = 0;
		while index < self.textToAnalyse.length {
			var range = NSRange(location: 0,length: 1)
			var value : AnyObject? = self.textToAnalyse.attribute(attribute, atIndex: index, effectiveRange: &range)
			if value != nil {
				characters.appendAttributedString(self.textToAnalyse.attributedSubstringFromRange(range))
				index = range.location + range.length
			}
			else {
				index++
			}
		}
		
		return characters
	}
	
}
